### PR TITLE
[fix](stream_load)fix bug when stream without content-length or chunk…

### LINKED ISSUE #27671: [Bug] be http server not ok
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

2.0.3

### What's Wrong?

be进程存在，且心跳正常，执行查询ok。但是http服务貌似停止了，无法正常采集metric
这里可以看到心跳正常
image
这里可以看到http拒绝服务
image
doris manager显示这个be无效了
image

### What You Expected?

正常采集metric

### How to Reproduce?

不知

### Anything Else?

无

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by xingyingone**
## Proposed changes

Issue Number: close https://github.com/apache/doris/issues/27671

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by xingyingone** [comment]
run buildall

File: be/src/http/action/stream_load.cpp [hunk 1]
```diff
@@ -301,6 +301,14 @@ Status StreamLoadAction::_on_header(HttpRequest* http_req, std::shared_ptr<Strea
         }
     }
 
+    if (ctx->body_bytes == 0 && !ctx->is_chunked_transfer) {
```
Comment (HHoflittlefish777) [inline be/src/http/action/stream_load.cpp:304]: Ctx->body_bytes == 0 must be zero in on_header method, for it hasn't started receive data here yet, which begins in StreamLoadAction::_handle. I think you can check in  _handle or use http_req->header(HttpHeaders::CONTENT_LENGTH).empty() to check.

File: be/src/http/action/stream_load.cpp [hunk 2]
```diff
@@ -301,6 +301,14 @@ Status StreamLoadAction::_on_header(HttpRequest* http_req, std::shared_ptr<Strea
...
+        LOG(WARNING) << "content_length is empty and transfer-encoding!=chunked, please set "
```
Comment (HHoflittlefish777) [inline be/src/http/action/stream_load.cpp:305]: It is no need log and return InternalError is enough.

**Comment by xingyingone** [comment]
run buildall

**Comment by xingyingone** [comment]
run buildall

File: be/src/http/action/stream_load.cpp (hunk 1 above)
Comment (HHoflittlefish777) [inline be/src/http/action/stream_load.cpp:304]: There may be several things to consider here:
1. Forgot to set content length or chunked
2. Incorrect 0 set
3. It is itself 0

**Review by freemandealer — COMMENTED** [review]
usually, when a function/branch is added, we need also to provide a UT or regression test. Please take a look at [here](https://doris.apache.org/zh-CN/community/developer-guide/regression-testing)

**Comment by xingyingone** [comment]
> usually, when a function/branch is added, we need also to provide a UT or regression test. Please take a look at [here](https://doris.apache.org/zh-CN/community/developer-guide/regression-testing)

Okay , I will provide it

**Comment by xingyingone** [comment]
run buildall

**Comment by xingyingone** [comment]
run buildall

File: regression-test/suites/load_p0/stream_load/test_stream_load.groovy [hunk 3]
```diff
@@ -1575,5 +1575,138 @@ suite("test_stream_load", "p0") {
...
+           DISTRIBUTED BY HASH(`k1`) BUCKETS 3
+           PROPERTIES ("replication_allocation" = "tag.location.default: 1");
+       """
+   
+       def command = "curl --location-trusted -u ${context.config.feHttpUser}:${context.config.feHttpPassword} -H column_separator:| -H ${db}:${tableName16} -H Content-Length:0  -H Transfer-Encoding:chunked -H columns:k1,k2,v1,v2,v3 -T ${context.dataPath}/test_chunked_transfer.csv http://${beHost}:${beHttpPort}/api/${db}/${tableName16}/_stream_load"
+       log.info("test chunked transfer command: ${command}")
+       def process = command.execute()
+       code = process.waitFor()
+       out = process.text
+       json2pc = parseJson(out)
```
Comment (HHoflittlefish777) [inline regression-test/suites/load_p0/stream_load/test_stream_load.groovy:1611]: Do not use name "json2pc", json enough

File: regression-test/suites/load_p0/stream_load/test_stream_load.groovy [hunk 4]
```diff
@@ -1575,5 +1575,138 @@ suite("test_stream_load", "p0") {
...
+        json2pc = parseJson(out)
+        log.info("test chunked transfer result: ${out}".toString())
+        def json = parseJson(out)
+        assertEquals("success", json.Status.toLowerCase())
+        assertEquals(0, json.NumberTotalRows)
+    } finally {
+        sql """ DROP TABLE IF EXISTS ${tableName16} FORCE"""
+    }
+
+
```
Comment (HHoflittlefish777) [inline regression-test/suites/load_p0/stream_load/test_stream_load.groovy:1710]: Can you add test with empty file?

**Comment by xingyingone** [comment]
run buildall

**Comment by xingyingone** [comment]
run buildall

**Comment by xingyingone** [comment]
run buildall

**Review by sollhui — APPROVED** [review]
LGTM

**Comment by xingyingone** [comment]
run buildall

**Review by alexxing662 — APPROVED** [review]

File: be/src/http/action/stream_load.cpp [hunk 5]
```diff
@@ -302,12 +309,21 @@ Status StreamLoadAction::_on_header(HttpRequest* http_req, std::shared_ptr<Strea
                 evhttp_request_get_connection(http_req->get_evhttp_request()), csv_max_body_bytes);
 #endif
     }
-
     if (!http_req->header(HttpHeaders::TRANSFER_ENCODING).empty()) {
         if (http_req->header(HttpHeaders::TRANSFER_ENCODING).find(CHUNK) != std::string::npos) {
             ctx->is_chunked_transfer = true;
         }
     }
+    if (http_req->header(HttpHeaders::CONTENT_LENGTH).empty() && !ctx->is_chunked_transfer) {
```
Comment (freemandealer) [inline be/src/http/action/stream_load.cpp:317]: please add UNLIKELY to make branch predictor happy

File: be/src/http/action/stream_load.cpp [hunk 6]
```diff
@@ -302,12 +309,21 @@ Status StreamLoadAction::_on_header(HttpRequest* http_req, std::shared_ptr<Strea
...
+        LOG(WARNING) << "content_length is empty and transfer-encoding!=chunked, please set "
+                        "content_length or transfer-encoding=chunked";
+        return Status::InternalError(
+                "content_length is empty and transfer-encoding!=chunked, please set content_length "
+                "or transfer-encoding=chunked");
+    } else if (!http_req->header(HttpHeaders::CONTENT_LENGTH).empty() && ctx->is_chunked_transfer) {
```
Comment (freemandealer) [inline be/src/http/action/stream_load.cpp:323]: ditto, UNLIKELY

File: regression-test/suites/load_p0/stream_load/test_stream_load.groovy [hunk 7]
```diff
@@ -1575,5 +1575,136 @@ suite("test_stream_load", "p0") {
...
+        """
+
+        def command = "curl --location-trusted -u ${context.config.feHttpUser}:${context.config.feHttpPassword} -H column_separator:| -H ${db}:${tableName16} -H Content-Length:0  -H Transfer-Encoding: -H columns:k1,k2,v1,v2,v3 -T ${context.dataPath}/test_chunked_transfer.csv http://${beHost}:${beHttpPort}/api/${db}/${tableName16}/_stream_load"
+        log.info("test chunked transfer command: ${command}")
+        def process = command.execute()
+        code = process.waitFor()
+        out = process.text
+        log.info("test chunked transfer result: ${out}".toString())
+        def json = parseJson(out)
+        assertEquals("success", json.Status.toLowerCase())
```
Comment (freemandealer) [inline regression-test/suites/load_p0/stream_load/test_stream_load.groovy:1673]: why success?

File: regression-test/suites/load_p0/stream_load/test_stream_load.groovy [hunk 8]
```diff
@@ -1575,5 +1575,136 @@ suite("test_stream_load", "p0") {
...
+        """
+
+        def command = "curl --location-trusted -u ${context.config.feHttpUser}:${context.config.feHttpPassword} -H column_separator:| -H ${db}:${tableName16} -H Content-Length:0  -H Transfer-Encoding:  -H columns:k1,k2,v1,v2,v3 -T ${context.dataPath}/test_chunked_transfer.csv http://${context.config.feHttpAddress}/api/${db}/${tableName16}/_stream_load"
+        log.info("test chunked transfer command: ${command}")
+        def process = command.execute()
+        code = process.waitFor()
+        out = process.text
+        log.info("test chunked transfer result: ${out}".toString())
+        def json = parseJson(out)
+        assertEquals("success", json.Status.toLowerCase())
```
Comment (freemandealer) [inline regression-test/suites/load_p0/stream_load/test_stream_load.groovy:1703]: why success?

File: be/src/http/action/stream_load.cpp [hunk 9]
```diff
@@ -144,6 +144,13 @@ Status StreamLoadAction::_handle(std::shared_ptr<StreamLoadContext> ctx) {
         LOG(WARNING) << "recevie body don't equal with body bytes, body_bytes=" << ctx->body_bytes
                      << ", receive_bytes=" << ctx->receive_bytes << ", id=" << ctx->id;
         return Status::InternalError("receive body don't equal with body bytes");
+    } else if (ctx->body_bytes == 0 && ctx->receive_bytes && ctx->is_chunked_transfer == false) {
```
Comment (freemandealer) [inline be/src/http/action/stream_load.cpp:147]: unlikely

**Comment by xingyingone** [comment]
run buildall

**Comment by xingyingone** [comment]
run buildall

**Comment by xingyingone** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**